### PR TITLE
norman: k-NN EdgeConv pre-encoder (local geometric feature aggregation before Transolver)

### PR DESCRIPTION
**Pull request by morganmcg1**
# norman: k-NN EdgeConv pre-encoder (local geometric feature aggregation before Transolver)

## Hypothesis

The Transolver processes point clouds via global slice-based attention, which may struggle to capture local geometric structure in the point cloud. A k-NN EdgeConv pre-encoder adds local neighborhood aggregation before the Transolver trunk: for each point, it aggregates edge features from its k nearest neighbors using a shared MLP. This gives the model explicit local geometric context (nearby surface curvature, local flow patterns) before the global attention.

EdgeConv (Wang et al., 2019 — DGCNN) computes `h_i = AGG_{j ∈ kNN(i)} MLP([h_i, h_j − h_i])`, where `h_j − h_i` encodes the relative displacement (edge features). Skip-connecting the output back to the input (initialized to zero weight) means training starts from the SOTA baseline, and the EdgeConv only contributes once it's learned a useful local aggregation.

## Implementation

This requires ~150 LOC change, primarily a new pre-encoder module and k-NN graph construction. The k-NN graph is best constructed in the DataLoader worker to avoid recomputation.

```python
# Pre-encoder module (~80 LOC)
class EdgeConvPreEncoder(nn.Module):
    """
    k-NN EdgeConv pre-encoder. 
    For each point i: h_i' = AGG_{j ∈ kNN(i)} MLP([h_i, h_j - h_i])
    Skip connection: output = input + EdgeConv(input)
    """
    def __init__(self, in_dim, k=16):
        super().__init__()
        self.k = k
        # MLP on edge features: input is [h_i (in_dim) || h_j-h_i (in_dim)] = 2*in_dim
        self.edge_mlp = nn.Sequential(
            nn.Linear(2 * in_dim, in_dim),
            nn.ReLU(),
            nn.Linear(in_dim, in_dim)
        )
        # Initialize last linear layer to zero for skip-connection safety
        nn.init.zeros_(self.edge_mlp[-1].weight)
        nn.init.zeros_(self.edge_mlp[-1].bias)
    
    def forward(self, x, edge_index):
        """
        x: (N, in_dim) — point features
        edge_index: (2, E) — k-NN graph edges (row=center, col=neighbor)
        Returns: (N, in_dim) — updated features (input + EdgeConv)
        """
        row, col = edge_index  # row=i (center), col=j (neighbor)
        
        # Edge features: [h_i, h_j - h_i]
        h_i = x[row]           # (E, in_dim)
        h_j = x[col]           # (E, in_dim)
        edge_feat = torch.cat([h_i, h_j - h_i], dim=-1)  # (E, 2*in_dim)
        
        # MLP on edge features
        edge_out = self.edge_mlp(edge_feat)  # (E, in_dim)
        
        # Aggregate (mean) over neighbors for each center point
        out = torch.zeros_like(x)
        out.scatter_add_(0, row.unsqueeze(1).expand_as(edge_out), edge_out)
        # Count neighbors per point for mean aggregation
        count = torch.zeros(x.shape[0], device=x.device)
        count.scatter_add_(0, row, torch.ones(row.shape[0], device=x.device))
        count = count.clamp(min=1).unsqueeze(1)
        out = out / count
        
        return x + out  # skip connection
```

**k-NN graph construction (in DataLoader or preprocessing):**

```python
# Build k-NN graph from coordinates using torch_cluster
from torch_cluster import knn_graph

# coords: (N, 3) — point coordinates
# Returns edge_index: (2, N*k) — directed graph (each point has k out-edges)
edge_index = knn_graph(coords, k=16, loop=False)
```

**Integration into model:**

```python
# In model __init__:
self.pre_encoder = EdgeConvPreEncoder(in_dim=input_feature_dim, k=16)

# In model forward, before Transolver trunk:
# surface_features: (N_surf, in_dim)
# surface_edge_index: (2, N_surf * k) — built from surface coords only
surface_features = self.pre_encoder(surface_features, surface_edge_index)
# Then pass augmented features to Transolver trunk as normal
```

**Important implementation notes:**
1. Build separate k-NN graphs for surface and volume points — do NOT mix them
2. k=16 strikes a good balance between local context and computation cost
3. The zero-init on the last MLP layer ensures skip-connection fallback to SOTA baseline
4. If `torch_cluster` is not available, implement k-NN manually with `torch.cdist` + `topk` (slower but no dependency)
5. EdgeConv hidden dim = input dim (no expansion) to avoid overfitting
6. The k-NN graph is fixed per sample (based on coordinates, not learned) — compute once per batch, not per epoch

## Training Command

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent norman \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --grad-clip-norm 0.5 \
  --pos-encoding-mode string_separable \
  --wandb-group norman-knn-edgeconv
```

Do not modify any other parameters beyond adding the EdgeConv pre-encoder code change.

## Baseline Metrics to Beat

**Current SOTA:** PR #311, W&B run `gcwx9yaa`, edward STRING-separable positional encoding

| Metric | Current SOTA (PR #311) | AB-UPT Reference | Gap (×) |
|--------|----------------------|------------------|---------|
| `val_abupt` (primary) | **7.546%** | — | — |
| `test_abupt` | **8.771%** | — | — |
| `surface_pressure` | 4.485% | 3.82% | ×1.17 |
| `wall_shear` | 8.227% | 7.29% | ×1.13 |
| `volume_pressure` | 12.438% | 6.08% | ×2.05 |
| `tau_x` | 7.253% | 5.35% | ×1.36 |
| `tau_y` | 9.233% | 3.65% | **×2.53** |
| `tau_z` | 10.449% | 3.63% | **×2.88** |

A merge requires: `val_abupt < 7.546%` AND `test_abupt < 8.771%`.

## Why This Should Work

- Transolver's global slice attention may miss local geometric features (sharp edges, boundary layer gradients)
- EdgeConv is the backbone of DGCNN, which excels at point cloud classification and segmentation — direct transfer to CFD surface point clouds
- The local neighborhood structure of aerodynamic surfaces is physically meaningful: neighboring points share similar boundary conditions
- Zero-init skip connection: training starts from SOTA baseline; the model only uses EdgeConv if it helps
- Expected impact: surface_pressure and wall_shear (local physical quantities that depend on nearby geometry)

## What to Report

1. Your rank-0 W&B run ID
2. Best val_abupt epoch and value
3. Final/best test metrics (all axes)
4. Whether surface_pressure and wall_shear improved (primary diagnostic — local quantities)
5. Any OOM or performance issues with EdgeConv computation (report k-NN build time per batch)

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Kickoff — clear hypothesis, be careful with implementation scope

Your EdgeConv pre-encoder hypothesis is well-motivated. The idea that global slice attention misses local geometric structure is correct — this is a known limitation of attention-based point cloud models, and DGCNN's EdgeConv is a proven fix.

**Key implementation checkpoints:**

1. **k-NN graph construction**: Build the k-NN graph once per sample (16 neighbors, coordinate space). Best done in the Dataset `__getitem__` so DataLoader workers parallelize it. Store as `edge_index` in the batch.

2. **Edge features**: `[h_i, h_j - h_i]` where `h_i` are the input point features (NOT raw coordinates). Starting from raw input features (7–8 channels) before the input projection means the MLP input is `2 * n_input_features`.

3. **Zero-init skip**: Your zero-init on the last linear is correct — this ensures training starts from the baseline. After a few epochs, EdgeConv should start contributing.

4. **Overhead**: k-NN graph construction for ~45k points (k=16) adds ~50ms per sample. With DDP on 8 GPUs, this should be masked by data loading if done in workers. Profile the first 3 steps and report data-loading time vs forward-pass time.

5. **Memory**: EdgeConv with k=16 on 45k points uses ~45k×16×2×8 = ~92M floats for edge features per batch sample — watch VRAM.

**Baseline target**: val_abupt < 7.546% (PR #311 benchmark). Current SOTA uses STRING-sep PE + grad-clip-norm 0.5.

**Reproduce command for your PR** (using your branch):
```bash
python train.py --pos-encoding-mode string_separable --grad-clip-norm 0.5 \
  --use-edgeconv-preencoder --edgeconv-k 16 \
  --wandb_group edgeconv_preencoder
```
(Adjust flag names to match your implementation.)

Post an intermediate ep3 or ep5 W&B link as soon as it's available — I want to see if EdgeConv is helping early trajectory vs the baseline.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #403 because Closing as misassignment — student 'norman' does not have a senpai-drivaerml-ddp8 GPU pod allocated for this research target. The DrivAerML pod fleet only includes: alphonse, askeladd, edward, fern, frieren, nezuko, tanjiro, thorfinn (8 students total, not the 16 across all targets). This PR cannot run on this branch.

The hypothesis itself is solid and may be re-assigned to one of the DrivAerML-allocated students in a future round. Apologies for the misroute — closing without prejudice to the underlying idea..
